src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs

```diff
@@ -0,0 +1,204 @@
+            {
+                return null;
+            }
+
+            var sb = new StringBuilder(1000);
+            var unmatchedCriteriaCount = _unmatchedCriteria.Count;
+            for (var i = 0; i < unmatchedCriteriaCount; i++)
+            {
+                sb.Append(_unmatchedCriteria[i]);
+                if (i + 1 < unmatchedCriteriaCount)
+                {
+                    sb.Append("\r\n");
+                }
+            }
+
+            return sb.ToString();
+        }
+
+        private List<string> _unmatchedCriteria = new(1000);
+        private List<string> _newUnmatchedCriteria = new(1000);
+
+        /**
+         * <summary>Called to tell this to persist the current frame as the last succesful key frame.  This is NOT done automatically in matched because we process multiple transient frames in a single pass and you don't want to update to a newer state yet when a later transient frame passes before the earlier one.</summary>
+         */
+        public void PersistPriorFrameStatus()
+        {
+            var uiTransforms = InGameObjectFinder.GetInstance().GetUITransformsForCurrentFrame();
+            var gameObjectTransforms = InGameObjectFinder.GetInstance().GetGameObjectTransformsForCurrentFrame();
+            // copy the dictionaries
+            _priorKeyFrameUIStatus = uiTransforms.Item2.ToDictionary(a => a.Key, a => a.Value);
+            _priorKeyFrameGameObjectStatus = gameObjectTransforms.Item2.ToDictionary(a=>a.Key, a=>a.Value);
+        }
+
+        /**
+         * <summary>Publicly callable.. caches the statuses of the last passed key frame for computing delta counts from</summary>
+         */
+        public bool Matched(KeyFrameCriteria[] criteriaList)
+        {
+            _newUnmatchedCriteria.Clear();
+            bool matched = MatchedHelper(AndOr.And, criteriaList);
+            if (matched)
+            {
+                _unmatchedCriteria.Clear();
+                _newUnmatchedCriteria.Clear();
+            }
+            else
+            {
+                (_unmatchedCriteria, _newUnmatchedCriteria) = (_newUnmatchedCriteria, _unmatchedCriteria);
+                _newUnmatchedCriteria.Clear();
+            }
+            return matched;
+        }
+
+        /**
+         * <summary>Only to be called internally by KeyFrameEvaluator</summary>
+         */
+        internal bool MatchedHelper(AndOr andOr, KeyFrameCriteria[] criteriaList)
+        {
+            var uiTransforms = InGameObjectFinder.GetInstance().GetUITransformsForCurrentFrame();
+            var gameObjectTransforms = InGameObjectFinder.GetInstance().GetGameObjectTransformsForCurrentFrame();
+
+            var normalizedPathsToMatch = new List<KeyFrameCriteria>();
+            var orsToMatch = new List<KeyFrameCriteria>();
+            var andsToMatch = new List<KeyFrameCriteria>();
+            //var pathsToMatch = new List<KeyFrameCriteria>();
+            //var xPathsToMatch = new List<KeyFrameCriteria>();
```

Comment:
I can remove, but 'future', I had comments like this throughout the code to help as 'reminders' while coding of what would come next and where it would fit